### ISSUE TITLE
may the host and schemes be optional?

### ISSUE THREAD

**Issue by dsilvasc** [twilio/guardrail#218]
I upgraded recently from guardrail 0.43.0 to 0.45.2, and in doing so the cli crashed:

```
Exception in thread "main" java.net.URISyntaxException: Expected scheme-specific part at index 5: http:
        at java.net.URI$Parser.fail(URI.java:2848)
        at java.net.URI$Parser.failExpecting(URI.java:2854)
        at java.net.URI$Parser.parse(URI.java:3057)
        at java.net.URI.<init>(URI.java:673)
        at com.twilio.guardrail.Common$.$anonfun$prepareDefinitions$8(Common.scala:55)
        at cats.data.NonEmptyList.map(NonEmptyList.scala:76)
        at com.twilio.guardrail.Common$.$anonfun$prepareDefinitions$7(Common.scala:53)
        at scala.Option.map(Option.scala:146)
        at com.twilio.guardrail.Common$.$anonfun$prepareDefinitions$1(Common.scala:53)
```

Looks like guardrail now expects `host` and `schemes` to be defined:
https://github.com/twilio/guardrail/blob/320afebc2c26f800de7699bd009b314d4ed4c6be/modules/codegen/src/main/scala/com/twilio/guardrail/Common.scala#L55

Are these required in swagger or can they be optional?

**Comment by blast-hardcheese** [comment]
My understanding is that URL should provide enough information on how to connect to a host. How is your URL configured?

**Comment by blast-hardcheese** [comment]
I will add, we did need to add some adapter code from OpenAPI 3.x models back to how we were doing things with Swagger, but it seems as though there's some missing test coverage here. If you could supply a failing specification, that would be appreciated

**Comment by dsilvasc** [comment]
In our swagger.yaml files we tend to omit both `host` and `schemes`.  The client fills them in (dev, prod, localhost, etc).

**Comment by blast-hardcheese** [comment]
That should be supported, can you provide a minimal reproducing case?

**Comment by dsilvasc** [comment]
example:

```yaml
---
swagger: "2.0"
info:
  version: "1.0.1"
basePath: "/v1"
paths:
  /foo/{bar}:
    get:
      operationId: "getFoo"
      produces:
      - "application/json"
      parameters:
      - name: "bar"
        in: "path"
        required: true
        type: "integer"
        pattern: "[0-9]+"
        format: "int64"
      responses:
        200:
          description: "successful operation"
```

Command line:

```bash
java -cp $GUARDRAIL_CLASSPATH com.twilio.guardrail.CLI --defaults --import scala.language.higherKinds --import akka.NotUsed --server --specPath src/main/resources/swagger-specs/swagger.yaml --outputPath build/generated --packageName repro
```

output:

```
Exception in thread "main" java.net.URISyntaxException: Expected scheme-specific part at index 5: http:
        at java.net.URI$Parser.fail(URI.java:2848)
        at java.net.URI$Parser.failExpecting(URI.java:2854)
        at java.net.URI$Parser.parse(URI.java:3057)
        at java.net.URI.<init>(URI.java:673)
        at com.twilio.guardrail.Common$.$anonfun$prepareDefinitions$8(Common.scala:55)
        at cats.data.NonEmptyList.map(NonEmptyList.scala:76)
        at com.twilio.guardrail.Common$.$anonfun$prepareDefinitions$7(Common.scala:53)
        at scala.Option.map(Option.scala:146)
        at com.twilio.guardrail.Common$.$anonfun$prepareDefinitions$1(Common.scala:53)
...
        at com.twilio.guardrail.CLICommon$.run(CLI.scala:80)
        at com.twilio.guardrail.CLICommon$$anonfun$handleLanguage$1.$anonfun$applyOrElse$2(CLI.scala:127)
        at com.twilio.guardrail.CLICommon$$anonfun$handleLanguage$1.$anonfun$applyOrElse$2$adapted(CLI.scala:127)
        at com.twilio.guardrail.CLICommon.main(CLI.scala:132)
        at com.twilio.guardrail.CLICommon.main$(CLI.scala:130)
        at com.twilio.guardrail.CLI$.main(CLI.scala:136)
        at com.twilio.guardrail.CLI.main(CLI.scala)
```

**Comment by blast-hardcheese** [comment]
That should work, judging by https://github.com/twilio/guardrail/blob/320afebc2c26f800de7699bd009b314d4ed4c6be/modules/codegen/src/main/scala/com/twilio/guardrail/Common.scala#L50

**Comment by dsilvasc** [comment]
`getUrl` maybe returns a string that's invalid when parsed by `new URI`

**Comment by blast-hardcheese** [comment]
@dsilvasc Would you please try out https://github.com/blast-hardcheese/guardrail/tree/issue-218?

```
sbt:guardrail-root> cli --defaults --import scala.language.higherKinds --import akka.NotUsed --server --specPath modules/sample/src/main/resources/issues/issue218.yaml --outputPath /tmp/blep/ --packageName repro --client --specPath modules/sample/src/main/resources/issues/issue218.yaml --outputPath /tmp/blep/ --packageName repro
[info] Running com.twilio.guardrail.CLI --defaults --import scala.language.higherKinds --import akka.NotUsed --server --specPath modules/sample/src/main/resources/issues/issue218.yaml --outputPath /tmp/blep/ --packageName repro --client --specPath modules/sample/src/main/resources/issues/issue218.yaml --outputPath /tmp/blep/ --packageName repro
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.

[success] Total time: 3 s, completed Mar 29, 2019 12:38:56 PM
sbt:guardrail-root>
```

**Comment by dsilvasc** [comment]
@blast-hardcheese Looks like that fixes it 👍

**Comment by blast-hardcheese** [comment]
@dsilvasc Excellent! Thanks for confirming!

**Comment by blast-hardcheese** [comment]
Resolved in 0.45.3, which has now been released. Feel free to reopen if necessary, as I deviated slightly from the version you tested, in order to maintain the default scheme (http) in the case of having the domain segment of the `host` specified.
